hls4ml/backends/vivado/passes/fifo_depth_optimization.py

```diff
@@ -0,0 +1,102 @@
+import json
+
+from pyDigitalWaveTools.vcd.parser import VcdParser
+
+import hls4ml
+from hls4ml.model.optimizer.optimizer import ConfigurableOptimizerPass, ModelOptimizerPass
+
+
+class FifoDepthOptimization(ConfigurableOptimizerPass, ModelOptimizerPass):
+    def __init__(self):
+        self.values = []
+
+    def match(self, node):
+        return True
+
+    def _populate_values(self, name, data, depth):
+        self.values.append({'name': name, 'data': [], 'max': 0, 'depth': 0})
+        get_values = lambda x: int(x[1][1:], 2)
+        self.values[-1]['data'] = [get_values(x) for x in data]
+        self.values[-1]['max'] = max(self.values[-1]['data'])
+        self.values[-1]['depth'] = int(depth[0][1][1:], 2)
+
+    def transform(self, model):
+        model.fifo_opt = True
+        init_large_fifo = getattr(self, 'init_large_fifo', True)
+        # cfg = model.config.config.copy()
+        # hls_config = cfg['HLSConfig']
+
+        # check axi-stream or io-stream, if not one the 2 exit
+        if not(model.config.get_config_value('IOType') == 'io_stream' or
+               model.config.get_config_value('AcceleratorConfig')['Interface'] == 'axi_stream' or
+               model.config.get_config_value('AcceleratorConfig')['Interface'] == 'axi_master'):
+            raise Exception('To use this optimization you have to set `IOType` field to `io_stream` in the HLS config '
+                            'or `axi_stream` or `axi_master` in `AcceleratorConfig` interface field')
+
+        # initialize all the fifos to 10000 so that they will be automatically implemented in BRAMs and so they will be
+        # profiled
+
+        if init_large_fifo:
+
+            for k, v in model.output_vars.items():
+                if model.config.get_config_value('Backend') == 'Vivado' and (v == model.get_input_variables()[0] or v == model.get_output_variables()[0]):
+                    continue
+                v.pragma = (v.pragma[0], 100000)
+
+        model.write()
+        model.build(reset=False, csim=True, synth=True, cosim=True, validation=False, export=False, vsynth=False)
+
+        with open(
+                model.config.get_output_dir() + '/' + model.config.get_project_name() + '_prj' + '/solution1/sim/verilog/fifo_opt.vcd') as vcd_file:
+            vcd = VcdParser()
+            vcd.parse(vcd_file)
+            data = vcd.scope.toJson()
+
+        # wrapper fifos - useful only with VivadoAccelerator backend
+        if model.config.get_config_value('Backend') == 'VivadoAccelerator':
```

Comment:
It would be nicer not to ask for the backend in a backend-registered optimizer pass. Could we register one pass for Vivado and another one for VivadoAccelerator, factorising those parts which are common?